src/JNI/com/amazonaws/kinesis/video/producer/jni/KinesisVideoClientWrapper.cpp

```diff
@@ -456,11 +456,86 @@ void KinesisVideoClientWrapper::putKinesisVideoFragmentMetadata(jlong streamHand
 
     if (STATUS_FAILED(retStatus))
     {
-        DLOGE("Failed to put a metadata with status code 0x%08x", retStatus);
-        throwNativeException(env, EXCEPTION_NAME, "Failed to put a metadata into the stream.", retStatus);
+        DLOGE("Failed to put a fragment metadata with status code 0x%08x", retStatus);
+        throwNativeException(env, EXCEPTION_NAME, "Failed to put a fragment metadata into the stream.", retStatus);
+        return;
+    }
+}
+
+void KinesisVideoClientWrapper::putKinesisVideoEventMetadata(jlong streamHandle, jint event, jobject kinesisStreamEventMetadata)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    JNIEnv *env;
+    mJvm->GetEnv((PVOID*) &env, JNI_VERSION_1_6);
+
+    if (!IS_VALID_CLIENT_HANDLE(mClientHandle))
+    {
+        DLOGE("Invalid client object");
+        throwNativeException(env, EXCEPTION_NAME, "Invalid call after the client is freed.", STATUS_INVALID_OPERATION);
         return;
     }
 
+    if (!IS_VALID_STREAM_HANDLE(streamHandle))
+    {
+        DLOGE("Invalid stream handle 0x%016" PRIx64, (UINT64) streamHandle);
+        throwNativeException(env, EXCEPTION_NAME, "Invalid stream handle.", STATUS_INVALID_OPERATION);
+        return;
+    }
+
+    // Check for a negative STREAM_EVENT_TYPE coming from Java.
+    if (event < STREAM_EVENT_TYPE_NONE) {
```

Comment:
Can we also check the upper limit here?

Reply:
No, we want to keep the upper limit check in PIC/ProducerC as we do for most of the existing parameters here in the JNI. However, we have to check for a negative value here as it is a signed int coming from Java.